File: src/chapters/nonlinear-equations/content/code/newton-nd.m
function [x, k] = newton_system(F, J, x0, tol, max_iter)
% NEWTON_SYSTEM  Newton's method for F(x)=0 with Jacobian J(x).
    if nargin < 4, tol = 1e-12; end
    if nargin < 5, max_iter = 100; end
    x = x0(:);
    for k = 1:max_iter
        Fx = F(x);
        if norm(Fx, inf) < tol, return; end
        x = x - J(x) \ Fx;
    end
end

% --- Demo ---
F = @(v) [v(1)^2 + v(2)^2 - 4; v(1)*v(2) - 1];
J = @(v) [2*v(1), 2*v(2); v(2), v(1)];
disp(newton_system(F, J, [2; 0.5])');

File: src/chapters/nonlinear-equations/content/code/newton-nd.py
import numpy as np


def newton_system(F, J, x0, tol=1e-12, max_iter=100):
    """Newton's method for F(x) = 0 with Jacobian J(x)."""
    x = np.array(x0, float)
    for k in range(1, max_iter + 1):
        Fx = np.array(F(x), float)
        if np.linalg.norm(Fx, np.inf) < tol:
            return x, k
        x = x - np.linalg.solve(np.array(J(x), float), Fx)
    return x, max_iter


if __name__ == "__main__":
    F = lambda v: [v[0]**2 + v[1]**2 - 4, v[0] * v[1] - 1]
    J = lambda v: [[2*v[0], 2*v[1]], [v[1], v[0]]]
    print(newton_system(F, J, [2.0, 0.5]))
